### ISSUE TITLE
Windows support for hooks

### ISSUE THREAD

**Issue by Alizter** [uptech/git-ps-rs#236]
When we get the Windows build a bit further, we should investigate how hooks can work on Windows and help support any missing features.

**Comment by drewdeponte** [comment]
@Alizter we now have builds working on Windows. The CI setup proves this. I am not sure what other features or support we would need. If you could test this out and research this further it would be great.

**Comment by Alizter** [comment]
My thinking with this issue is that the hooks we recommend for users are shell scripts which are not straightforward to run on windows. There are definitely ways to achieve this however it will need to be looked into.

**Comment by drewdeponte** [comment]
Yeah I would imagine based on my old school windows knowledge that they could just implement them as bat files.

Technically they can be any executable format. So they could even be executable binaries. 

But yeah I would be good for someone who uses windows to play around with this so we can provide direction in the docs.

**Comment by Alizter** [comment]
One thing to try would be a .bat file like such:
```bat
@echo off

:: isolate_post_checkout hook using GitHub CLI (https://cli.github.com)
::
:: This hook handles verifying a patch is isolated after checking out the
:: temporary ps/tmp/isolate branch.
::
:: gps exposes the remote name and remote url respective as arg 1 and 2
::
:: Generally you want this script to verify that your code builds,
:: your tests pass, and any other constraints you want to add here
:: to make sure your code is good prior to integration. People often add linting

set remote_name=%1
set remote_url=%2

cargo build
set BUILD_EXIT_CODE=%ERRORLEVEL%
if not %BUILD_EXIT_CODE% equ 0 (
  echo "cargo build failed with exit code %BUILD_EXIT_CODE%"
  exit /b 1
)

cargo test -q
set TEST_EXIT_CODE=%ERRORLEVEL%
if not %TEST_EXIT_CODE% equ 0 (
  echo "cargo test failed with exit code %TEST_EXIT_CODE%"
  exit /b 2
)

exit /b 0
```

**Comment by drewdeponte** [comment]
@Alizter tested this and seems we run into a problem on windows where .bat files, etc. aren't flagged as executable in terms of the file system and currently we have a pre-check for execution flag so that we can provide better feedback to users and abort the commands earlier on. Rather than running everything and then failing out when the execution of the hook fails because the execution flag wasn't set. This makes a bunch of sense on unix platforms but seems to make less sense on Windows.

If I recall correctly from my discussion with @Alizter the current thinking is that we should make the hook commands a configuration. That way on windows users could define them to be something that would execute a `.bat` file using the bat interpreter executable, etc.

This would provide more flexibility than we currently have and would enable support of windows. The only downside is that we would have to get rid of that pre-check as it is no longer about a file existing that is an executable as an entire command is given to us. Maybe this is just a loss we have to accept and think about with each command to see if there is anything we can do on a command by command basis to attempt to address this.

**Comment by drewdeponte** [comment]
We should look at how git does this and probably follow suite

**Comment by drewdeponte** [comment]
Ok, so did some digging into Git and how it supports hooks. It looks like on Windows Git supports hooks through Cygwin and Bash.

From everything I have read you need to make sure that the NTFS permissions are set correctly to executable for the file. I am not sure how to do that as I don't have a windows machine.

Here is a website that somewhat explains it https://www.tygertec.com/git-hooks-practical-uses-windows/

Anyways. I feel like since `gps` is an extension on top of Git. If anyone is seriously using it on Windows then they will have gotten hooks working with Git proper. So hooks working the same way in `gps` shouldn't be a problem.

@Alizter can you test this as you were testing things before with respect to Windows?

This is a link where someone sets up a hook to have a Git hook on Windows in turn execute a Powershell script. From what I found this is a common thing to do on Windows because Bash is quite limited in functionality on Windows while Powershell is not.

https://blog.nimblepros.com/blogs/setting-up-pre-commit-git-hook-on-windows-with-powershell/

**Comment by judemille** [comment]
The way Git for Windows handles this is rather intricate.
From the [start_command](https://github.com/git-for-windows/git/blob/main/run-command.c#L930) function, it calls into [mingw_spawnvpe](https://github.com/git-for-windows/git/blob/main/compat/mingw.c#L2259), seemingly a support harness not actually in MinGW.

Eventually, it checks for the existence of either the base file or the file with `.exe` at the end, potentially on `PATH`, and if it doesn't have `.exe` on the end, it runs an [interpreter check](https://github.com/git-for-windows/git/blob/main/compat/mingw.c#L1695) on it.

I would love to not have to reinvent the wheel here, but I don't know of any good way to just use these functions from Rust.
